patching/EAP7-1063_EAP_Patching_With_Galleon.adoc

```diff
@@ -0,0 +1,121 @@
+
+==== Provisioning state history
+
+Galleon saves a description (in terms of feature-pack configurations) of the current provisioning state before every provisioning operation (such as install, uninstall, provision state, update) into the state history which allows the user to undo the last operation go back to the previous state (undo requires access to the artifact repository as well). A user can limit or disable the state history.
+
+==== User changes and provisioning state transition
+
+Before installing a new feature-pack (patch, version update, new software) into existing installation, Galleon will identify user changes since the last provisioned state. That includes changes to the configuration and other files. The changes are calculated as a diff between the last provisioned state and the current state.
+
+Next, Galleon provisions an updated installation at a staged location and if that went ok, it applies the user changes identified previously to the newly provisioned state.
+
+User changes to the files that haven't changed in the updated installation are re-applied to the updated installation.
+
+If a user changed a file in the installation that has changed in the updated installation then if the changes can be merged (text files, property files) they are merged. If the changes can't be merged, the file with the user changes will appear next to the updated file with a suffix ".prev".
+
+Standalone configuration changes are identified as diffs in terms of feature configurations and then re-applied to the corresponding updated configurations.
+
+Host and domain configuration changes are not currently merged as the standalone ones and *not* merged as text files. If user modified host or domain configurations and the corresponding configurations have changed in the updated installation, the user-modified configuration files will appear in the installation with suffix ".prev".
+
+
+== Issue Metadata
+
+=== Issue:
+
+* {issue-base-url}EAP7-1063[EAP7-1063]
+
+=== Related Issues:
+
+* {issue-base-url}WFCORE-3932[WFCORE-3932]
+
+=== Dev Contacts:
+
+* mailto:alexey.loubyansky@redhat.com[Alexey Loubyansky]
+
+=== QE Contacts:
+
+* mailto:rjanik@redhat.com[Richard Janík]
+
+=== Affected Projects or Components:
+
+Any project using the tool or its API for patching and/or provisioning is going to be affected.
+
+== Requirements
+
+=== Hard requirements
+
+* Install feature-packs, including patches (directly as a file and/or from a repository providing coordinates or using the discovery mechanism);
+* Uninstall installed feature-packs, including patches;
+* Maintain provisioning state history with the ability to undo going back to previously provisioned state;
+* Preserve user changes during patching and version updates.
+
+=== Soft requirements
+
+== Limitations
+
+Requirement to have access to the repository (local or remote).
+
```

Comment:
In the limitations section we need a discussion of whether the server can be / cannot be running when its FS is being patched.

I'm pretty sure it's "cannot be". Which is an improvement disguised as a limitation, as trying to patch a running server was probably not necessary yet is the source of many of the problems of the current patching.

Related: not being able to patch a running server would disallow using the HAL tool to apply patches.

Reply:
Another limitation that follows from not allowing a running server is that remote patching is not supported. The galleon tool is designed to make local FS changes; it is not sending commands to some remote process that is making the FS changes.

Reply:
Right. We could imagine developing an agent for both use-cases above. For the running server though, it would be picking the time to shut it down and reboot the updated version.